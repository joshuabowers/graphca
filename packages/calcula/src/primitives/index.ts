import { isPrimitive, PrimitiveNode } from '../closures/primitive'
import { Real, real, isReal, EulerMascheroni } from './real'
import { 
  Complex, complex, isComplex, ComplexInfinity, isComplexInfinity 
} from './complex'
import { Boolean, boolean, isBoolean } from './boolean'
import { Nil, nil, isNil } from './nil'
import { NaN, nan, isNaN } from './nan'

export type Primitive = Real | Complex | Boolean | Nil | NaN
export { Real, Complex, Boolean, Nil, NaN }

export type PrimitiveFns = 
| typeof real | typeof complex | typeof boolean

export { real, complex, boolean, nil, nan }
export { isReal, isComplex, isBoolean, isNil, isNaN, isComplexInfinity }
export { isPrimitive, PrimitiveNode }
export { EulerMascheroni, ComplexInfinity }
