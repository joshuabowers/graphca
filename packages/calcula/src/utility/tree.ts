import { Writer } from "../monads/writer"

export enum Clades {
  primitive = 'Primitive',  // Constants: 10, 4 - 5i, true
  unary = 'Unary',          // Functions and Ops: not, abs, sin
  binary = 'Binary',        // Functions and Ops: and, log, polygamma
  variadic = 'Variadic'     // Variables: x, OhGodWhy
}

export enum Genera {
  arithmetic = 'Arithmetic',
  connective = 'Connective',
  inequalities = 'Inequalities',
  trigonometric = 'Trigonometric',
  arcus = 'Arcus',
  hyperbolic = 'Hyperbolic',
  areaHyperbolic = 'AreaHyperbolic',
  combinatorics = 'Combinatorics'
}

export enum Species {
  // Primitives
  real = 'Real',
  complex = 'Complex',
  boolean = 'Boolean',
  nil = 'Nil',
  nan = 'NaN',

  // Arithmetic
  add = 'Addition',
  multiply = 'Multiplication',
  raise = 'Exponentiation',

  // Functions
  abs = 'Absolute',

  // :: Inequalities
  equals = 'Equality',
  notEquals = 'Inequality',
  lessThan = 'LessThan',
  greaterThan = 'GreaterThan',
  lessThanEquals = 'LessThanEquals',
  greaterThanEquals = 'GreaterThanEquals',

  // :: Connectives
  not = 'Complement',
  and = 'Conjunction',
  or = 'Disjunction',
  xor = 'ExclusiveDisjunction',
  implies = 'Implication',
  nand = 'AlternativeDenial',
  nor = 'JointDenial',
  xnor = 'Biconditional',
  converse = 'ConverseImplication',

  // :: Trigonometric
  sin = 'Sine',
  cos = 'Cosine',
  tan = 'Tangent',
  csc = 'Cosecant',
  sec = 'Secant',
  cot = 'Cotangent',

  // Variables
  variable = 'Variable'
}

export type TreeNode = {
  readonly clade: Clades,
  readonly genus?: Genera,
  readonly species: Species
}

export type DerivedNode<T extends ((...args: any[]) => any)> = 
  ReturnType<T> extends Writer<infer U> ? U : never

export type TreeNodeGuardFn<T extends TreeNode> = 
  (value: Writer<TreeNode>) => value is Writer<T>

// values should be Writer<TreeNode>, narrowed to Writer<T>
export const isClade = <T extends TreeNode>(clade: Clades) =>
  (value: Writer<TreeNode>): value is Writer<T> =>
    value?.value.clade === clade

export const isGenus = <T extends TreeNode>(genus: Genera) =>
  (value: Writer<TreeNode>): value is Writer<T> =>
    value?.value.genus === genus

export const isSpecies = <T extends TreeNode>(species: Species) =>
  (value: Writer<TreeNode>): value is Writer<T> =>
    value?.value.species === species

export const notAny = (...args: Species[]) => {
  const exclude = new Set(args)
  return <T extends TreeNode>(t: Writer<T>) => 
    t.value && !exclude.has(t.value.species)
}

export const any = (...args: Species[]) => {
  const include = new Set(args)
  return <T extends TreeNode>(t: Writer<T>) => 
    t.value && include.has(t.value.species)
}
