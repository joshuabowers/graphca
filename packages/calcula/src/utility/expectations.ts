import { method, multi, Multi, _ } from '@arrows/multimethod'
import { isWriter, Writer } from '../monads/writer'
import { TreeNode } from './tree'
import { Real, Complex, isReal, isComplex } from '../primitives'

export type ExpectCloseTo = Multi 
  & ((actual: Writer<Real>, expected: Writer<Real>, precision: number) => void)
  & ((actual: Writer<Complex>, expected: Writer<Complex>, precision: number) => void)

export const expectCloseTo: ExpectCloseTo = multi(
  method(
    [isReal, isReal, _], 
    (actual: Writer<Real>, expected: Writer<Real>, precision: number) => 
      expect(actual.value.value).toBeCloseTo(expected.value.value, precision)
  ),
  method(
    [isComplex, isComplex, _],
    (actual: Writer<Complex>, expected: Writer<Complex>, precision: number) => {
      expect(actual.value.a).toBeCloseTo(expected.value.a, precision)
      expect(actual.value.b).toBeCloseTo(expected.value.b, precision)
    }
  )
)

type Input = TreeNode|Writer<TreeNode>|[Input, Input]
export type Operation = [Input, string]

export const expectWriter = <
  Actual extends TreeNode, 
  Expected extends TreeNode|Writer<TreeNode>
>(
  actual: Writer<Actual>
) => (expected: Expected, ...operations: Operation[]) => {
  expect(actual).toEqual({
    value: isWriter(expected) ? expected.value : expected,
    log: operations.map(([input, action]) => ({
      input: isWriter(input) 
        ? input.value 
        : (Array.isArray(input) 
          ? input.map(i => isWriter(i) ? i.value : i)
          : input), 
      action
    }))
  })
}
