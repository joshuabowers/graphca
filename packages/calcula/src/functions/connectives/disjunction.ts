import { _ } from '@arrows/multimethod'
import { TreeNode, Genera, Species, Notation } from "../../utility/tree"
import { Boolean, boolean } from "../../primitives"
import { binary, when } from "../../closures/binary"
import { deepEquals, isValue } from "../../utility/deepEquals"
import { Connective } from './connective'
import { Complement, isComplement } from './complement'
import { isConjunction } from './conjunction'
import { implies } from './implication'
import { converse } from './converseImplication'
import { Unicode } from '../../Unicode'

export type Disjunction = Connective<Species.or>

export const [or, isDisjunction, $or] = binary<Disjunction, Boolean>(
  Unicode.or, Notation.infix, Species.or, Genera.connective
)(
  (l, r) => boolean(l.value.value !== 0 || r.value.value !== 0),
  (l, r) => boolean(
    l.value.a !== 0 || l.value.b !== 0 || r.value.a !== 0 || r.value.b !== 0
  ),
  (l, r) => boolean(l.value.value || r.value.value)
)(
  when(
    [_, isValue(boolean(false))],
    (l, _r) => [l, 'disjunctive identity']
  ),
  when(
    [isValue(boolean(false)), _],
    (_l, r) => [r, 'disjunctive identity']
  ),
  when(
    [_, isValue(boolean(true))],
    [boolean(true), 'disjunctive annihilator']
  ),
  when(
    [isValue(boolean(true)), _],
    [boolean(true), 'disjunctive annihilator']
  ),
  when(
    deepEquals, 
    (l, _r) => [l, 'disjunctive idempotency']
  ),
  when(
    (l, r) => isConjunction(r) && deepEquals(l, r.value.left),
    (l, _r) => [l, 'disjunctive absorption']
  ),
  when(
    (l, r) => isConjunction(r) && deepEquals(l, r.value.right),
    (l, _r) => [l, 'disjunctive absorption']
  ),
  when(
    (l, r) => isConjunction(l) && deepEquals(l.value.left, r),
    (_l, r) => [r, 'disjunctive absorption']
  ),
  when(
    (l, r) => isConjunction(l) && deepEquals(l.value.right, r),
    (_l, r) => [r, 'disjunctive absorption']
  ),
  when(
    (l, r) => isComplement(r) && deepEquals(l, r.value.expression),
    [boolean(true), 'disjunctive tautology']
  ),
  when(
    (l, r) => isComplement(l) && deepEquals(l.value.expression, r),
    [boolean(true), 'disjunctive tautology']
  ),
  when<Complement, TreeNode>(
    (l, _r) => isComplement(l),
    (l, r) => [
      implies(l.value.expression, r), 
      'disjunctive implication'
    ]
  ),
  when<TreeNode, Complement>(
    (_l, r) => isComplement(r),
    (l, r) => [
      converse(l, r.value.expression), 
      'disjunctive converse implication'
    ]
  )
)
