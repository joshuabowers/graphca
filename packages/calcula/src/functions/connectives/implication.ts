import { _ } from '@arrows/multimethod'
import { Genera, Species, Notation } from "../../utility/tree"
import { Boolean, boolean } from "../../primitives"
import { binary, when } from "../../closures/binary"
import { isValue } from "../../utility/deepEquals"
import { Connective } from './connective'
import { not } from './complement'
import { or } from './disjunction'
import { Unicode } from '../../Unicode'

export type Implication = Connective<Species.implies>

export const [implies, isImplication, $implies] = binary<Implication, Boolean>(
  Unicode.implies, Notation.infix, Species.implies, Genera.connective
)(
  (l, r) => or(not(l), r),
  (l, r) => or(not(l), r),
  (l, r) => or(not(l), r)
)(
  when(
    [isValue(boolean(true)), _], 
    (_l, r) => [r, 'implicative identity']
  ),
  when(
    [_, isValue(boolean(true))], 
    [boolean(true), 'implicative annihilator']
  ),
  when(
    [isValue(boolean(false)), _], 
    [boolean(true), 'implicative annihilator']
  ),
  when(
    [_, isValue(boolean(false))], 
    (l, _r) => [
      not(l), 'implicative complementation'
    ]
  )
)
