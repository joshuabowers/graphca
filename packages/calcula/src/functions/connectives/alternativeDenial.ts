import { _ } from '@arrows/multimethod'
import { Writer, unit } from "../../monads/writer"
import { TreeNode, Clades, Genera, Species, isSpecies, isGenus } from "../../utility/tree"
import { Real, Complex, Boolean, boolean } from "../../primitives"
import { BinaryNode, binary, when } from "../../closures/binary"
import { deepEquals, isValue } from "../../utility/deepEquals"
import { Connective } from './connective'
import { Complement, isComplement, not } from './complement'
import { and } from './conjunction'
import { or } from './disjunction'
import { implies } from './implication'

export type AlternativeDenial = Connective<Species.nand>

export const [nand, isAlternativeDenial, $nand] = binary<AlternativeDenial, Boolean>(
  Species.nand, Genera.connective
)(
  (l, r) => [not(and(unit(l), unit(r))), 'real alternative denial'],
  (l, r) => [not(and(unit(l), unit(r))), 'complex alternative denial'],
  (l, r) => [not(and(unit(l), unit(r))), 'boolean alternative denial']
)(
  when(
    [_, isValue(boolean(true))],
    (l, _r) => [not(unit(l)), 'alternative deniable complementation']
  ),
  when(
    [isValue(boolean(true)), _],
    (_l, r) => [not(unit(r)), 'alternative deniable complementation']
  ),
  when(
    [_, isValue(boolean(false))],
    [boolean(true), 'alternative deniable annihilator']
  ),
  when(
    [isValue(boolean(false)), _],
    [boolean(true), 'alternative deniable annihilator']
  ),
  when(
    deepEquals, (l, _r) => [not(unit(l)), 'alternative deniable complementation']
  ),
  when<TreeNode, AlternativeDenial>(
    (l, r) => isAlternativeDenial(r) && deepEquals(l, r.value.left),
    (l, r) => [implies(unit(l), r.right), 'alternative deniable implication']
  ),
  when<TreeNode, AlternativeDenial>(
    (l, r) => isAlternativeDenial(r) && deepEquals(l, r.value.right),
    (l, r) => [implies(unit(l), r.left), 'alternative deniable implication']
  ),
  when<Complement, Complement>(
    (l, r) => isComplement(l) && isComplement(r),
    (l, r) => [or(l.expression, r.expression), 'De Morgan']
  ),
  when(
    (l, r) => isComplement(r) && deepEquals(l, r.value.expression),
    [boolean(true), 'alternative deniable annihilator']
  ),
  when(
    (l, r) => isComplement(l) && deepEquals(l.value.expression, r),
    [boolean(true), 'alternative deniable annihilator']
  )
)
