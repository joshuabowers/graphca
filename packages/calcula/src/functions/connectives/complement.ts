import { _ } from '@arrows/multimethod'
import { Genera, Species, Notation, isSpecies } from "../../utility/tree"
import { Boolean, boolean } from "../../primitives"
import { Unary, unary, unaryFnRule, when } from "../../closures/unary"
import { Conjunction, and, andRule } from './conjunction'
import { Disjunction, or, orRule } from './disjunction'
import { ExclusiveDisjunction, xor, xorRule } from './exclusiveDisjunction'
import { Implication } from './implication'
import { AlternativeDenial, nand, nandRule } from './alternativeDenial'
import { JointDenial, nor, norRule } from './jointDenial'
import { Biconditional, xnor, xnorRule } from './biconditional'
import { ConverseImplication } from './converseImplication'
import { Unicode } from '../../Unicode'
import { rule } from '../../utility/rule'

export type Complement = Unary<Species.not, Genera.connective>

export const notRule = unaryFnRule(Unicode.not)

export const [not, isComplement, $not] = unary<Complement, Boolean>(
  Unicode.not, Notation.prefix, Species.not, Genera.connective
)(
  r => [boolean(r.value === 0), notRule(r), 'real complement'],
  c => [boolean(c.a === 0 && c.b === 0), notRule(c), 'complex complement'],
  b => [boolean(!b.value), notRule(b), 'boolean complement']
)(
  // NOTE: Cannot use derived isSpecies guards as they are not defined
  // by this point.
  when<Complement>(
    isSpecies(Species.not), 
    v => [v.expression, rule`${v.expression}`, 'double complement']
  ),
  when<Conjunction>(
    isSpecies(Species.and), 
    v => [nand(v.left, v.right), nandRule(v.left, v.right), 'complement of conjunction']
  ),
  when<Disjunction>(
    isSpecies(Species.or), 
    v => [nor(v.left, v.right), norRule(v.left, v.right), 'complement of disjunction']
  ),
  when<AlternativeDenial>(
    isSpecies(Species.nand), 
    v => [and(v.left, v.right), andRule(v.left, v.right), 'complement of alternative denial']
  ),
  when<JointDenial>(
    isSpecies(Species.nor), 
    v => [or(v.left, v.right), orRule(v.left, v.right), 'complement of joint denial']
  ),
  when<ExclusiveDisjunction>(
    isSpecies(Species.xor), 
    v => [xnor(v.left, v.right), xnorRule(v.left, v.right), 'complement of exclusive disjunction']
  ),
  when<Implication>(
    isSpecies(Species.implies),
    v => [
      and(v.left, not(v.right)), 
      rule`${v.left} ${Unicode.and} (${v.right})${Unicode.not}`, 
      'complement of implication'
    ]
  ),
  when<Biconditional>(
    isSpecies(Species.xnor),
    v => [xor(v.left, v.right), xorRule(v.left, v.right), 'complement of biconditional']
  ),
  when<ConverseImplication>(
    isSpecies(Species.converse),
    v => [
      and(not(v.left), v.right), 
      rule`(${v.left})${Unicode.not} ${Unicode.and} ${v.right}`,
      'complement of converse implication'
    ]
  )
)
