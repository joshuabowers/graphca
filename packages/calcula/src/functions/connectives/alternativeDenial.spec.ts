import { unit } from '../../monads/writer'
import { expectToEqualWithSnapshot } from '../../utility/expectations'
import { Clades, Genera, Species } from '../../utility/tree'
import { boolean } from '../../primitives'
import { variable } from '../../variable'
import { not } from './complement'
import { or } from './disjunction'
import { implies } from './implication'
import { nand, $nand } from './alternativeDenial'

describe('$nand', () => {
  it('generates an AlternativeDenial for a pair of TreeNode inputs', () => {
    expect(
      $nand(unit(variable('x').value), unit(variable('y').value))[0]
    ).toEqual({
      clade: Clades.binary, genus: Genera.connective, species: Species.nand,
      left: unit(variable('x').value), right: unit(variable('y').value)
    })
  })
})

describe('nand', () => {
  it('returns false when given two true things', () => {
    expectToEqualWithSnapshot(
      nand(boolean(true), boolean(true)),
      boolean(false)
    )
  })

  it('returns true if the left argument is false', () => {
    expectToEqualWithSnapshot(
      nand(boolean(false), boolean(true)),
      boolean(true)
    )
  })

  it('returns true if the right argument is false', () => {
    expectToEqualWithSnapshot(
      nand(boolean(true), boolean(false)),
      boolean(true)
    )
  })

  it('returns true if both arguments are false', () => {
    expectToEqualWithSnapshot(
      nand(boolean(false), boolean(false)),
      boolean(true)
    )
  })

  it('returns the complement of left operand if the right is true', () => {
    expectToEqualWithSnapshot(
      nand(variable('x'), boolean(true)),
      not(variable('x'))
    )
  })

  it('returns the right operand if the left is true', () => {
    expectToEqualWithSnapshot(
      nand(boolean(true), variable('x')),
      not(variable('x'))
    )
  })

  it('returns true if the right operand is false', () => {
    expectToEqualWithSnapshot(
      nand(variable('x'), boolean(false)),
      boolean(true)
    )
  })

  it('returns true if the left operand is false', () => {
    expectToEqualWithSnapshot(
      nand(boolean(false), variable('x')),
      boolean(true)
    )
  })

  it('returns the complement of the left operand if left equivalent to right', () => {
    expectToEqualWithSnapshot(
      nand(variable('x'), variable('x')),
      not(variable('x'))
    )
  })

  it('returns an implication if the right operand is a nand of the left', () => {
    expectToEqualWithSnapshot(
      nand(variable('x'), nand(variable('x'), variable('y'))),
      implies(variable('x'), variable('y'))
    )
    expectToEqualWithSnapshot(
      nand(variable('x'), nand(variable('y'), variable('x'))),
      implies(variable('x'), variable('y'))
    )
  })

  it('returns a disjunction of mutually complemented operands', () => {
    expectToEqualWithSnapshot(
      nand(not(variable('x')), not(variable('y'))),
      or(variable('x'), variable('y'))
    )
  })

  it('returns true if the right operand is the complement of the left', () => {
    expectToEqualWithSnapshot(
      nand(variable('x'), not(variable('x'))),
      boolean(true)
    )
  })

  it('returns true if the left operand is the complement of the right', () => {
    expectToEqualWithSnapshot(
      nand(not(variable('x')), variable('x')),
      boolean(true)
    )
  })

  it('returns a AlternativeDenial on variable input', () => {
    expectToEqualWithSnapshot(
      nand(variable('x'), variable('y')),
      $nand(variable('x'), variable('y'))[0]
    )
  })
})
