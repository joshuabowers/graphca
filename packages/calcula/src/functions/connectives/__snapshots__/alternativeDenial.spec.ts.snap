// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nand returns a AlternativeDenial on variable input 1`] = `
[
  [
    "(x↑y)",
    "identified alternative denial",
  ],
  [
    "([x]↑y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}↑[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x↑{y})",
    "processed right operand",
  ],
  [
    "(x↑y)",
    "created alternative denial",
  ],
]
`;

exports[`nand returns a disjunction of mutually complemented operands 1`] = `
[
  [
    "(¬(x)↑¬(y))",
    "identified alternative denial",
  ],
  [
    "([¬(x)]↑¬(y))",
    "processing left operand",
  ],
  [
    "¬(x)",
    "identified complement",
  ],
  [
    "¬([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "¬({x})",
    "processed argument",
  ],
  [
    "¬(x)",
    "created complement",
  ],
  [
    "({¬(x)}↑[¬(y)])",
    "processed left operand; processing right operand",
  ],
  [
    "¬(y)",
    "identified complement",
  ],
  [
    "¬([y])",
    "processing argument",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "¬({y})",
    "processed argument",
  ],
  [
    "¬(y)",
    "created complement",
  ],
  [
    "(¬(x)↑{¬(y)})",
    "processed right operand",
  ],
  [
    "(¬(x)↑¬(y))",
    "De Morgan",
  ],
  [
    "(x∨y)",
    "identified disjunction",
  ],
  [
    "([x]∨y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∨[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x∨{y})",
    "processed right operand",
  ],
  [
    "(x∨y)",
    "created disjunction",
  ],
]
`;

exports[`nand returns an implication if the right operand is a nand of the left 1`] = `
[
  [
    "(x↑(x↑y))",
    "identified alternative denial",
  ],
  [
    "([x]↑(x↑y))",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}↑[(x↑y)])",
    "processed left operand; processing right operand",
  ],
  [
    "(x↑y)",
    "identified alternative denial",
  ],
  [
    "([x]↑y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}↑[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x↑{y})",
    "processed right operand",
  ],
  [
    "(x↑y)",
    "created alternative denial",
  ],
  [
    "(x↑{(x↑y)})",
    "processed right operand",
  ],
  [
    "(x↑(x↑y))",
    "alternative deniable implication",
  ],
  [
    "(x→y)",
    "identified implication",
  ],
  [
    "([x]→y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}→[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x→{y})",
    "processed right operand",
  ],
  [
    "(x→y)",
    "created implication",
  ],
]
`;

exports[`nand returns an implication if the right operand is a nand of the left 2`] = `
[
  [
    "(x↑(y↑x))",
    "identified alternative denial",
  ],
  [
    "([x]↑(y↑x))",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}↑[(y↑x)])",
    "processed left operand; processing right operand",
  ],
  [
    "(y↑x)",
    "identified alternative denial",
  ],
  [
    "([y]↑x)",
    "processing left operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "({y}↑[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(y↑{x})",
    "processed right operand",
  ],
  [
    "(y↑x)",
    "created alternative denial",
  ],
  [
    "(x↑{(y↑x)})",
    "processed right operand",
  ],
  [
    "(x↑(y↑x))",
    "alternative deniable implication",
  ],
  [
    "(x→y)",
    "identified implication",
  ],
  [
    "([x]→y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}→[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x→{y})",
    "processed right operand",
  ],
  [
    "(x→y)",
    "created implication",
  ],
]
`;

exports[`nand returns false when given two true things 1`] = `
[
  [
    "(true↑true)",
    "identified alternative denial",
  ],
  [
    "([true]↑true)",
    "processing left operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "({true}↑[true])",
    "processed left operand; processing right operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "(true↑{true})",
    "processed right operand",
  ],
  [
    "(true↑true)",
    "alternative deniable complementation",
  ],
  [
    "¬(true)",
    "identified complement",
  ],
  [
    "¬([true])",
    "processing argument",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "¬({true})",
    "processed argument",
  ],
  [
    "¬(true)",
    "boolean complement",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`nand returns the complement of left operand if the right is true 1`] = `
[
  [
    "(x↑true)",
    "identified alternative denial",
  ],
  [
    "([x]↑true)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}↑[true])",
    "processed left operand; processing right operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "(x↑{true})",
    "processed right operand",
  ],
  [
    "(x↑true)",
    "alternative deniable complementation",
  ],
  [
    "¬(x)",
    "identified complement",
  ],
  [
    "¬([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "¬({x})",
    "processed argument",
  ],
  [
    "¬(x)",
    "created complement",
  ],
]
`;

exports[`nand returns the complement of the left operand if left equivalent to right 1`] = `
[
  [
    "(x↑x)",
    "identified alternative denial",
  ],
  [
    "([x]↑x)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}↑[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(x↑{x})",
    "processed right operand",
  ],
  [
    "(x↑x)",
    "alternative deniable complementation",
  ],
  [
    "¬(x)",
    "identified complement",
  ],
  [
    "¬([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "¬({x})",
    "processed argument",
  ],
  [
    "¬(x)",
    "created complement",
  ],
]
`;

exports[`nand returns the right operand if the left is true 1`] = `
[
  [
    "(true↑x)",
    "identified alternative denial",
  ],
  [
    "([true]↑x)",
    "processing left operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "({true}↑[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(true↑{x})",
    "processed right operand",
  ],
  [
    "(true↑x)",
    "alternative deniable complementation",
  ],
  [
    "¬(x)",
    "identified complement",
  ],
  [
    "¬([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "¬({x})",
    "processed argument",
  ],
  [
    "¬(x)",
    "created complement",
  ],
]
`;

exports[`nand returns true if both arguments are false 1`] = `
[
  [
    "(false↑false)",
    "identified alternative denial",
  ],
  [
    "([false]↑false)",
    "processing left operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "({false}↑[false])",
    "processed left operand; processing right operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "(false↑{false})",
    "processed right operand",
  ],
  [
    "(false↑false)",
    "alternative deniable annihilator",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`nand returns true if the left argument is false 1`] = `
[
  [
    "(false↑true)",
    "identified alternative denial",
  ],
  [
    "([false]↑true)",
    "processing left operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "({false}↑[true])",
    "processed left operand; processing right operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "(false↑{true})",
    "processed right operand",
  ],
  [
    "(false↑true)",
    "alternative deniable complementation",
  ],
  [
    "¬(false)",
    "identified complement",
  ],
  [
    "¬([false])",
    "processing argument",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "¬({false})",
    "processed argument",
  ],
  [
    "¬(false)",
    "boolean complement",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`nand returns true if the left operand is false 1`] = `
[
  [
    "(false↑x)",
    "identified alternative denial",
  ],
  [
    "([false]↑x)",
    "processing left operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "({false}↑[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(false↑{x})",
    "processed right operand",
  ],
  [
    "(false↑x)",
    "alternative deniable annihilator",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`nand returns true if the left operand is the complement of the right 1`] = `
[
  [
    "(¬(x)↑x)",
    "identified alternative denial",
  ],
  [
    "([¬(x)]↑x)",
    "processing left operand",
  ],
  [
    "¬(x)",
    "identified complement",
  ],
  [
    "¬([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "¬({x})",
    "processed argument",
  ],
  [
    "¬(x)",
    "created complement",
  ],
  [
    "({¬(x)}↑[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(¬(x)↑{x})",
    "processed right operand",
  ],
  [
    "(¬(x)↑x)",
    "alternative deniable annihilator",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`nand returns true if the right argument is false 1`] = `
[
  [
    "(true↑false)",
    "identified alternative denial",
  ],
  [
    "([true]↑false)",
    "processing left operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "({true}↑[false])",
    "processed left operand; processing right operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "(true↑{false})",
    "processed right operand",
  ],
  [
    "(true↑false)",
    "alternative deniable complementation",
  ],
  [
    "¬(false)",
    "identified complement",
  ],
  [
    "¬([false])",
    "processing argument",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "¬({false})",
    "processed argument",
  ],
  [
    "¬(false)",
    "boolean complement",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`nand returns true if the right operand is false 1`] = `
[
  [
    "(x↑false)",
    "identified alternative denial",
  ],
  [
    "([x]↑false)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}↑[false])",
    "processed left operand; processing right operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "(x↑{false})",
    "processed right operand",
  ],
  [
    "(x↑false)",
    "alternative deniable annihilator",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`nand returns true if the right operand is the complement of the left 1`] = `
[
  [
    "(x↑¬(x))",
    "identified alternative denial",
  ],
  [
    "([x]↑¬(x))",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}↑[¬(x)])",
    "processed left operand; processing right operand",
  ],
  [
    "¬(x)",
    "identified complement",
  ],
  [
    "¬([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "¬({x})",
    "processed argument",
  ],
  [
    "¬(x)",
    "created complement",
  ],
  [
    "(x↑{¬(x)})",
    "processed right operand",
  ],
  [
    "(x↑¬(x))",
    "alternative deniable annihilator",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;
